Plugins/NativeEngine/Source/VertexArray.cpp

```diff
@@ -35,44 +35,54 @@ namespace Babylon
         m_indexBufferRecord.Buffer = indexBuffer;
     }
 
-    void VertexArray::RecordVertexBuffer(VertexBuffer* vertexBuffer, uint32_t location, uint32_t byteOffset, uint32_t byteStride, uint32_t numElements, uint32_t type, bool normalized)
+    void VertexArray::RecordVertexBuffer(VertexBuffer* vertexBuffer, uint32_t location, uint32_t byteOffset, uint32_t byteStride, uint32_t numElements, uint32_t type, bool normalized, uint32_t divisor)
     {
-        bgfx::VertexLayout layout{};
-        layout.begin();
-
         auto attrib{static_cast<bgfx::Attrib::Enum>(location)};
-        auto attribType{static_cast<bgfx::AttribType::Enum>(type)};
-
-        // clang-format off
-        const bool promoteToFloats = !normalized
-            && (bgfx::getCaps()->rendererType == bgfx::RendererType::Direct3D11 ||
-                bgfx::getCaps()->rendererType == bgfx::RendererType::Direct3D12 ||
-                bgfx::getCaps()->rendererType == bgfx::RendererType::Vulkan)
-            && (attribType == bgfx::AttribType::Int8 ||
-                attribType == bgfx::AttribType::Uint8 ||
-                attribType == bgfx::AttribType::Uint10 ||
-                attribType == bgfx::AttribType::Int16 ||
-                attribType == bgfx::AttribType::Uint16);
-        // clang-format on
-
-        if (promoteToFloats)
+        if (divisor == 1)
         {
-            layout.add(attrib, static_cast<uint8_t>(numElements), bgfx::AttribType::Float);
-            layout.m_stride = static_cast<uint16_t>(sizeof(float) * numElements);
-            vertexBuffer->PromoteToFloats(attribType, numElements, byteOffset, byteStride);
+            // bgfx allows instancing on 5 vec4 attributes
+            if (m_vertexBufferInstanceRecords.size() < 5)
+            {
+                m_vertexBufferInstanceRecords[attrib] = {vertexBuffer, byteOffset, byteStride, static_cast<uint16_t>(sizeof(float) * numElements)};
+            }
```

Comment:
Probably should at least log a warning if more than 5?

Reply:
logged with `JsConsoleLogger`